src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -177,124 +196,169 @@ export interface SavedObjectsSetupDeps {
+    }
+
+    this.logger.debug('Starting SavedObjects service');
 
     const kibanaConfig = await this.coreContext.configService
       .atPath<KibanaConfigType>('kibana')
       .pipe(first())
       .toPromise();
-
     const savedObjectsConfig = await this.coreContext.configService
       .atPath<SavedObjectsConfigType>('migrations')
       .pipe(first())
       .toPromise();
+    const adminClient = this.setupDeps!.elasticsearch.adminClient;
+    const migrator = this.createMigrator(kibanaConfig, savedObjectsConfig, migrationsRetryDelay);
 
-    const adminClient = setupDeps.elasticsearch.adminClient;
-
-    const migrator = (this.migrator = new KibanaMigrator({
-      savedObjectSchemas,
-      savedObjectMappings,
-      savedObjectMigrations,
-      savedObjectValidations,
-      logger: this.coreContext.logger.get('migrations'),
-      kibanaVersion: this.coreContext.env.packageInfo.version,
-      config: setupDeps.legacyPlugins.pluginExtendedConfig,
-      savedObjectsConfig,
-      kibanaConfig,
-      callCluster: migrationsRetryCallCluster(
-        adminClient.callAsInternalUser,
-        this.coreContext.logger.get('migrations'),
-        migrationsRetryDelay
-      ),
-    }));
+    /**
+     * Note: We want to ensure that migrations have completed before
+     * continuing with further Core start steps that might use SavedObjects
+     * such as running the legacy server, legacy plugins and allowing incoming
+     * HTTP requests.
+     *
+     * However, our build system optimize step and some tests depend on the
+     * HTTP server running without an Elasticsearch server being available.
+     * So, when the `migrations.skip` is true, we skip migrations altogether.
+     */
+    const cliArgs = this.coreContext.env.cliArgs;
+    const skipMigrations = cliArgs.optimize || savedObjectsConfig.skip;
+    await migrator.runMigrations(skipMigrations);
 
-    const createSORepository = (callCluster: APICaller, extraTypes: string[] = []) => {
+    const createRepository = (callCluster: APICaller, extraTypes: string[] = []) => {
       return SavedObjectsRepository.createRepository(
         migrator,
-        savedObjectSchemas,
-        setupDeps.legacyPlugins.pluginExtendedConfig,
+        new SavedObjectsSchema(this.schemas),
+        this.setupDeps!.legacyPlugins.pluginExtendedConfig,
         kibanaConfig.index,
         callCluster,
         extraTypes
       );
     };
 
-    this.clientProvider = new SavedObjectsClientProvider<KibanaRequest>({
+    const repositoryFactory: SavedObjectsRepositoryFactory = {
+      createInternalRepository: (extraTypes?: string[]) =>
+        createRepository!(adminClient.callAsInternalUser, extraTypes),
+      createScopedRepository: (req: KibanaRequest, extraTypes?: string[]) =>
+        createRepository!(adminClient.asScoped(req).callAsCurrentUser, extraTypes),
+    };
+
+    const clientProvider = new SavedObjectsClientProvider<KibanaRequest>({
```

Comment:
I guess the intention was to allow the legacy SO service to scope it to raw `Request`, however it seems it's never actually used, as the legacy service use directly the NP one

https://github.com/elastic/kibana/blob/ca554024966ecca9a461fe53a61042317816dee0/src/legacy/server/saved_objects/saved_objects_mixin.js#L131-L143

The generic type can probably be removed.

Reply:
Yes, we can remove the generic type. The types are currently actually wrong because depending on where a request comes from (NP or legacy) the wrappers registered with the SavedObjectsClientProvider can either get a KibanaRequest or Request.

Reply:
> The types are currently actually wrong because depending on where a request comes from (NP or legacy) the wrappers registered with the SavedObjectsClientProvider can either get a KibanaRequest or Request.

I just saw that, by removing the generic type (that was defaulting to unknown in legacy) and forcing it to `KibanaRequest`. This generates a few TS errors in legacy/shim code, where an actual `Request` was provided, such as:

https://github.com/elastic/kibana/blob/ea9a7b8a16abbee97ddf39f91e4cf3c3b7253b42/x-pack/legacy/plugins/alerting/server/plugin.ts#L131-L137

I see two options:

- Change SavedObjectsClientWrapperOptions (and all other classes in `src/core/server/saved_objects/service/lib/scoped_client_provider.ts`) to remove the generic and use `KibanaRequest | Request`

```
export interface SavedObjectsClientWrapperOptions {
  client: SavedObjectsClientContract;
  request: KibanaRequest | Request;
}
```

- Do the same, but restrict to only `KibanaRequest` and uses `KibanaRequest.from(request)` in the plugins code.

I think the second one is more futur-proof, but WDYT?

Reply:
> Do the same, but restrict to only KibanaRequest and uses KibanaRequest.from(request) in the plugins code.

👍

Reply:
Done in https://github.com/elastic/kibana/pull/55012/commits/511c4fa59d67fad5a566990d8a60a6acbd7df02f